subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java

```diff
@@ -122,78 +106,53 @@ public void triggerWatchProbe(String path) {
             .forEach(probedHierarchy -> {
-                stopWatchingProbeForHierarchy(probedHierarchy);
-                probeRegistry.disarmWatchProbe(probedHierarchy);
+                File probeDirectory = probeRegistry.getProbeDirectory(probedHierarchy);
+                disarmWatchProbeForHierarchy(probedHierarchy, probeDirectory);
             });
 
-        if (!hierarchiesToStopWatching.isEmpty()) {
-            stopWatchingHierarchies(hierarchiesToStopWatching);
-        }
-        if (!hierarchiesToStartWatching.isEmpty()) {
-            hierarchiesToStartWatching.forEach(locationToWatchValidator::validateLocationToWatch);
-            startWatchingHierarchies(hierarchiesToStartWatching);
-        }
-
         probedHierarchies.stream()
             .filter(newProbedHierarchy -> !oldProbedHierarchies.contains(newProbedHierarchy))
             .forEach(probedHierarchy -> {
-                startWatchingProbeForHierarchy(probedHierarchy);
-                probeRegistry.armWatchProbe(probedHierarchy);
+                File probeDirectory = probeRegistry.getProbeDirectory(probedHierarchy);
+                // Make sure the directory exists, this can be necessary when
+                // included builds are evaluated with configuration cache
+                //noinspection ResultOfMethodCallIgnored
+                probeDirectory.mkdirs();
+                armWatchProbeForHierarchy(probedHierarchy, probeDirectory);
             });
-
-        LOGGER.info("Watching {} directory hierarchies to track changes", watchedHierarchies.size());
     }
 
-    protected abstract void startWatchingHierarchies(Collection<File> hierarchiesToWatch);
-
-    protected abstract void startWatchingProbeForHierarchy(File hierarchyToWatch);
-
-    protected abstract void stopWatchingHierarchies(Collection<File> hierarchiesToWatch);
-
-    protected abstract void stopWatchingProbeForHierarchy(File hierarchyToWatch);
+    protected FileHierarchySet updateWatchedHierarchies(SnapshotHierarchy root) {
+        return resolveWatchedFiles(watchableHierarchies, root);
+    }
 
-    public interface FileSystemLocationToWatchValidator {
-        FileSystemLocationToWatchValidator NO_VALIDATION = location -> {
-        };
+    @OverridingMethodsMustInvokeSuper
+    protected void armWatchProbeForHierarchy(File probedHierarchy, File probeDirectory) {
+        probeRegistry.armWatchProbe(probedHierarchy);
+    }
 
-        void validateLocationToWatch(File location);
+    @OverridingMethodsMustInvokeSuper
+    protected void disarmWatchProbeForHierarchy(File probedHierarchy, File probeDirectory) {
+        probeRegistry.disarmWatchProbe(probedHierarchy);
     }
 
     @VisibleForTesting
-    static FileHierarchySet resolveWatchedFiles(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot) {
+    protected static FileHierarchySet resolveWatchedFiles(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot) {
```

Comment:
Why is this now `protected` and not package private any more?

Reply:
No reason, reverted.